Extension metadata and generation of extensions.json

aloubyansky opened the pull request:
Fixes #833, #4317 

* Enhanced the quarkus-bootstrap:extension-descriptor mojo to also generate META-INF/quarkus-extension.json with the extension descr/props;
* Added quarkus:generate-extensions-json mojo that generates extensions.json file (and attaches it as a Maven artifact to the project) for a BOM by collecting all the META-INF/quarkus-extension.json files from its managed deps;
* Added devtools/extensions-json module that generates extensions.json file for Quarkus Core (which for now isn't used yet as it needs refinement).

I decided not to re-use `META-INF/quarkus-extension.properties` for this, I think we should deprecate it and have a simpler `META-INF/quarkus-extension-deployment-gav` or something to make it easier/faster to resolve the deployment artifact at build time.

FYI @maxandersen, @paulrobinson, @quintesse, @gastaldi

aloubyansky committed:
* Enhanced the quarkus-bootstrap:extension-descriptor mojo to also generate META-INF/quarkus-extension.json with the extension descr/props;
* Added quarkus:generate-extensions-json mojo that generates extensions.json file (and attaches it as a Maven artifact to the project) for a BOM by collecting all the META-INF/quarkus-extension.json files from its managed deps;
* Added devtools/core-extensions-json module that generates extensions.json file for Quarkus Core (which for now isn't used yet as it needs refinement).

gastaldi commented on devtools/maven/src/main/java/io/quarkus/maven/GenerateExtensionsJsonMojo.java:
```diff
@@ -0,0 +1,258 @@
+            }
+        }
+        @formatter:on
+        */
```
Remove this commented out code

gastaldi commented on independent-projects/bootstrap/pom.xml:
```diff
@@ -24,6 +24,7 @@
         <maven.compiler.source>1.8</maven.compiler.source>
         <maven.compiler.target>1.8</maven.compiler.target>
         <!-- Versions -->
+        <javax.json.version>1.1.4</javax.json.version>
```
Not required, but if you want a JSON Writer with less dependencies (specially on the javax.json API), there is https://github.com/ralfstx/minimal-json

gastaldi reviewed (approved):
LGTM, added a few minor comments

maxandersen commented:
Weren't we going to put these in quarkus-extension.properties rather than introduce another marker file ?

aloubyansky commented:
@maxandersen i mentioned my reasons in the summary of the PR.

aloubyansky commented:
I added a few changed since @gastaldi has approved it, specifically I added the BOM GAV from which the JSON was generated to the resulting JSON. So now it's an object with attributes: `bom` and `extensions` which is an array of extensions with the usual attributes.
So now this PR needs a new approval to get merged.

gsmet commented on extensions/agroal/runtime/src/main/resources/META-INF/quarkus-extension.json:
```diff
@@ -0,0 +1,8 @@
+
+{
+    "name": "Agroal - Database connection pool",
+    "labels": [
```
I'm wondering if we should include at least part of the info we have on the `extensions.yaml` file from the website?

See https://github.com/quarkusio/quarkusio.github.io/blob/develop/_data/extensions.yaml .

I'm thinking about:
- the description
- a key defining the category (I would keep the list of category key <-> name mapping and the order in the website - or on `code.quarkus.io`).

If we do that, I wouldn't have to add the extensions on the website.

Note that I haven't followed all the discussions so maybe it's a stupid idea :).

aloubyansky replied:
Sure. Here I'm taking what's in the current `extensions.json` as a start.

maxandersen replied:
+1 on getting the minimal version in first.
Next step there efter is to add the additional features such as description, category, etc.

aloubyansky commented:
Please, don't merge it yet. I'm looking into a build issue in Eclipse.

maxandersen commented:
> I decided not to re-use `META-INF/quarkus-extension.properties` for this, I think we should deprecate it and have a simpler `META-INF/quarkus-extension-deployment-gav` or something to make it easier/faster to resolve the deployment artifact at build time.

I'm not following why we wouldn't want to reuse the existing properites, nor why there should be `-gav` info at the end. Can you elaborate ?

aloubyansky commented:
The properties file has been used so far to store the GAV of the corresponding deployment artifact. While I did have in mind to possibly extend this file with more info (primarily provisioning related), the way it is evolving now is not looking to be effective to me.
1. We need to discover what the corresponding deployment artifact coords are during the build and test setup in the most efficient manner. Checking for a specific JAR entry containing only what we need it is a lot more efficient then loading a potentially large properties file looking for a single key.
2. It seems like what we need to have at the end is content in a JSON format, it is easier to provide it in this format in the first place and avoid transforming it into properties syntax and then back.

aloubyansky commented:
> Please, don't merge it yet. I'm looking into a build issue in Eclipse.

That's related to mojo parameter initialization. E.g. we've had a parameter `deployment` with some `defaultValue` properly initialized. If I add another parameter initialized the same way with the same value, this parameter will remain `null` when Eclipse is building the workspace. Yet, the code that I change in the plugin is executed, so it's not like it's picking the wrong impl of the plugin. Perhaps, the wrong metadata somehow. Perhaps, it's also related to the fact that I'm testing a SNAPSHOT version. This is too weird, I'm done with it.
The conclusion is building from Eclipse is not reliable. Not a surprise, I guess.

maxandersen commented:
It is intentional we are committing the marker files for each extension or do we plan to have our own extension using the mojo's ? 

also if we are introducing `META-INF/quarkus-extension.json` as the  official marker file I do think it would be better to either do `META-INF/quarkus-extension.yaml` or at least support both formats.

This was one of the reasons I was liking more `META-INF/quarkus-extension.properties` to start :)

aloubyansky commented:
> It is intentional we are committing the marker files for each extension or do we plan to have our own extension using the mojo's ?

Not sure I understand your question. Every extension is supposed to include the description. Whether it's generated or written maintained manually.

> also if we are introducing `META-INF/quarkus-extension.json` as the official marker file I do think it would be better to either do `META-INF/quarkus-extension.yaml` or at least support both formats.

Why?

> This was one of the reasons I was liking more `META-INF/quarkus-extension.properties` to start :)

To later support 3 formats?

maxandersen commented:
> Not sure I understand your question. Every extension is supposed to include the description. Whether it's generated or written maintained manually.

just curious as thought you said you wanted the extensions in core to have the maven plugin generate it.

maxandersen commented:
> > also if we are introducing `META-INF/quarkus-extension.json` as the official marker file I do think it would be better to either do `META-INF/quarkus-extension.yaml` or at least support both formats.
> 
> Why?

yaml is better for humans  - json is for machines ;)

maxandersen commented:
> > This was one of the reasons I was liking more `META-INF/quarkus-extension.properties` to start :)
> 
> To later support 3 formats?

no, to avoid having to deal with  anything but built-in to java parsing/formats :)

maxandersen commented on devtools/maven/src/main/java/io/quarkus/maven/GenerateExtensionsJsonMojo.java:
```diff
@@ -0,0 +1,203 @@
+                    processMetaInfDir(artifact, artifactFs.getPath(BootstrapConstants.META_INF), extJsonBuilder);
+                }
+            }
+        } catch (Throwable t) {
```
shouldn't this not just catch (Exception e) ?

aloubyansky replied:
Thanks, I am actually throwing a wrong exception here. Copy-pasted from another class.

maxandersen commented on devtools/maven/pom.xml:
```diff
@@ -100,6 +100,11 @@
         </dependency>
 
+        <dependency>
+            <groupId>org.glassfish</groupId>
```
would this not be better/sufficient:

```
<groupId>javax.json</groupId>
<artifactId>javax.json-api<artifactId>
```

maxandersen replied:
ah - gdammit - keep forgetting that detail ;)

maxandersen commented on devtools/maven/src/main/java/io/quarkus/maven/GenerateExtensionsJsonMojo.java:
```diff
@@ -0,0 +1,203 @@
+        if (!Files.exists(metaInfDir)) {
+            return false;
+        }
+        final Path p = metaInfDir.resolve(BootstrapConstants.EXTENSION_PROPS_JSON_FILE_NAME);
```
I'm getting a compile error on this:

```
ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0-jboss-2:compile (default-compile) on project quarkus-maven-plugin: Compilation failure
[ERROR] /Users/max/code/quarkus/quarkus/devtools/maven/src/main/java/io/quarkus/maven/GenerateExtensionsJsonMojo.java:[172,61] cannot find symbol
[ERROR]   symbol:   variable EXTENSION_PROPS_JSON_FILE_NAME
[ERROR]   location: interface io.quarkus.bootstrap.BootstrapConstants
```

missing a commit/add of `BootstrapConstants.java` ?

maxandersen replied:
never mind - mistakenly only compiled devtools ;)

maxandersen commented on independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java:
```diff
@@ -18,14 +19,22 @@
  */
-@Mojo(name = "extension-descriptor", defaultPhase = LifecyclePhase.COMPILE, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME)
+@Mojo(name = "extension-descriptor", defaultPhase = LifecyclePhase.PROCESS_RESOURCES, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME)
 public class ExtensionDescriptorMojo extends AbstractMojo {
```
is it expected that on first build you get errors like:

```
ERROR] Could not find goal 'generate-extensions-json' in plugin io.quarkus:quarkus-maven-plugin:999-SNAPSHOT among available goals add-extension, add-extensions, analyze-call-tree, build, create, create-extension, dev, generate-config, help, list-extensions, native-image, remote-dev -> [Help 1]
``` ?

second build works since then the plugin is available to use.

its as if the sequence of builds are not right.

not a biggie but just worth mentioning in case it can be avoided.

aloubyansky replied:
I couldn't reproduce it neither. It might have used a locally available version of the plugin that didn't include this mojo. If you remove quarkus artifacts from your local maven repo and rebuild you shouldn't see this issue.

maxandersen replied:
i'll assume I had a bad state as I can't reproduce it neither now :)

maxandersen reviewed (approved):
There are few minor things in this but I approve this as its generally what we want to get to and I would rather have generation of extensions.json in core now than wait.

The other things we can fix in follow-up PR's as we haven't yet removed core extensions.json thus not breaking anything.

maxandersen commented:
just noticed that the generated quarkus-extension.json in some cases ends up with:

```
{
  "groupId": null,
  "artifactId": null
}
```
i.e.  `./vertx-core/runtime/target/classes/META-INF/quarkus-extension.json`

others have no GAV at all:
`./jgit/runtime/src/main/resources/META-INF/quarkus-extension.json`:

```
{
    "name": "JGit",
    "labels": [
        "git"
    ]
}
```

and finally also when existing data you get:

`./flyway/runtime/target/classes/META-INF/quarkus-extension.json`

```
{
  "name": "Flyway",
  "labels": [
    "flyway",
    "database",
    "data"
  ],
  "guide": "https://quarkus.io/guides/flyway-guide",
  "groupId": null,
  "artifactId": null
}
```

Shouldn't there at least be a groupid and artifactid in here ?

gsmet commented:
@maxandersen I wasn't able to reproduce your last issue with groupId and artifactId being null. Everything is OK here.

gsmet reviewed (commented):
I added a few questions.

It needs a rebase and once it's done, if you could run a full build with an empty repo and check that:
* it works
* we don't have these null groupIds and artifactIds issues

Once you have checked that, I would say let's commit it and make incremental changes on top of it.

maxandersen commented:
> I wasn't able to reproduce your last issue with groupId and artifactId being null. Everything is OK here.

shouldn't `mvn package` be sufficient ?

gsmet commented:
Ah, that's a good question for Alexey. With Quarkus, I usually do `mvn clean install`

maxandersen commented:
I did a new run with `mvn clean package` and now I have groupId/artifactId filled out where they before was `null`.

Still have neither in ` ./jgit/runtime/src/main/resources/META-INF/quarkus-extension.json`:
```
{
    "name": "JGit",
    "labels": [
        "git"
    ]
}
```

aloubyansky commented:
> > > also if we are introducing `META-INF/quarkus-extension.json` as the official marker file I do think it would be better to either do `META-INF/quarkus-extension.yaml` or at least support both formats.
> > 
> > 
> > Why?
> yaml is better for humans - json is for machines ;)

I can do that. But I'll keep only one format. The aggregated extensions list will still be in JSON though, I guess.

aloubyansky commented on independent-projects/bootstrap/maven-plugin/pom.xml:
```diff
@@ -61,6 +61,10 @@
         </dependency>
+        <dependency>
+            <groupId>com.eclipsesource.minimal-json</groupId>
+            <artifactId>minimal-json</artifactId>
```
I was originally using the glassfish one but then George suggested me to use this impl as a smaller and faster one. Given that it's an independent project, I didn't mind changing. I won't mind changing back neither.

aloubyansky commented:
> just noticed that the generated quarkus-extension.json in some cases ends up with:
> 
> ```
> {
>   "groupId": null,
>   "artifactId": null
> }
> ```
> 
> i.e. `./vertx-core/runtime/target/classes/META-INF/quarkus-extension.json`
> 
> others have no GAV at all:
> `./jgit/runtime/src/main/resources/META-INF/quarkus-extension.json`:
> 
> ```
> {
>     "name": "JGit",
>     "labels": [
>         "git"
>     ]
> }
> ```
> 
> and finally also when existing data you get:
> 
> `./flyway/runtime/target/classes/META-INF/quarkus-extension.json`
> 
> ```
> {
>   "name": "Flyway",
>   "labels": [
>     "flyway",
>     "database",
>     "data"
>   ],
>   "guide": "https://quarkus.io/guides/flyway-guide",
>   "groupId": null,
>   "artifactId": null
> }
> ```
> 
> Shouldn't there at least be a groupid and artifactid in here ?

That's probably your IDE building it? There is a comment somewhere above where I summarized my struggling figuring out why Eclipse (in my case) didn't properly initialize Mojo parameters.

aloubyansky commented:
> I did a new run with `mvn clean package` and now I have groupId/artifactId filled out where they before was `null`.
> 
> Still have neither in ` ./jgit/runtime/src/main/resources/META-INF/quarkus-extension.json`:
> 
> ```
> {
>     "name": "JGit",
>     "labels": [
>         "git"
>     ]
> }
> ```

Yes, but that's `src` dir. It's a template. You should be checking the `target` one.

gsmet commented:
The `target` one is perfectly OK for me.

aloubyansky commented:
> > I wasn't able to reproduce your last issue with groupId and artifactId being null. Everything is OK here.
> 
> shouldn't `mvn package` be sufficient ?

Worked for me.

maxandersen commented:
> That's probably your IDE building it? There is a comment somewhere above where I summarized my struggling figuring out why Eclipse (in my case) didn't properly initialize Mojo parameters.

could be - I had code open that uses m2e behind the scenes.
Lets not block this on that issue but I opened https://github.com/quarkusio/quarkus/issues/4542 as it would affect extension writers so would be good to have a way to handle/workaround this.

FroMage commented:
Damnit, I only noticed this issue now. I would prefer to generate that `quarkus-extension.json` file from our existing maven extension plugin if the following properties are defined (and the file does not exist already):

```
<project>
 ...
 <properties>
  <quarkus.extension.name>RESTEasy - Jackson</quarkus.extension.name>
  <quarkus.extension.labels>resteasy-jackson,jaxrs-json,resteasy-json,resteasy,jaxrs,json,jackson</quarkus.extension.labels>
 </properties>
...
</project>
```

FroMage commented on independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java:
```diff
@@ -78,5 +95,37 @@ public void execute() throws MojoExecutionException {
         } catch(IOException e) {
             throw new MojoExecutionException("Failed to persist extension descriptor " + output.resolve(BootstrapConstants.DESCRIPTOR_FILE_NAME), e);
         }
+
```
@aloubyansky I don't understand this code: it appears to read the file if it exists, and then writes it back. Surely I'm reading this wrong?

aloubyansky replied:
Not entirely wrong :) At this point, the file that's actually on the disk is missing groupId and artifactId, which makes it more like a template that needs to be complete by the plugin. However, you're right, it could originally be complete, in which case persisting it is useless.

FroMage replied:
But you're reading it from the output folder. Shouldn't it be read from the src folder?

aloubyansky replied:
I am reading it from the output folder to support filtering of resources, just in case.

FroMage replied:
OK, if it's done on purpose then fine :)

ia3andy commented:
> Please, don't merge it yet. I'm looking into a build issue in Eclipse.

We should introduce the WIP bot in here?

maxandersen commented:
> We should introduce the WIP bot in here?

isn't that superseeded by github now directly supporting drafts pull-requests ?
